### ISSUE TITLE
Bloomberg feedback on TypeScript 4.7 Beta

### ISSUE THREAD

**Issue by dragomirtitian** [microsoft/TypeScript#49071]
We are in the process of evaluating the impact of upgrading to 4.7 on our codebase. Here some preliminary findings:

| #  | Change  |  Affects  | Release notes  | Packages affected  | Introduced in |
|---|---|---|---|---|---|
| 1 | Detection of weak generic types | Type checking | Announced | 100% | [#48366](https://github.com/microsoft/TypeScript/pull/48366) | 
| 2 | Symbol included in `keyof T` | Type checking | Not Announced | 2% |   | 
| 3 | Better always true detection | Type checking | Not Announced | <1% |   |
| 4 | Comments on default exports have moved | d.ts emit | Not Announced | 5%
| 5 | Optional parameter does not on include undefined | d.ts emit | Not Announced | 2% | [#48605](https://github.com/microsoft/TypeScript/pull/48605)  |
| 6 | Changes in emitted conditional types | d.ts emit | Not Announced | <1% | [#48592](https://github.com/microsoft/TypeScript/pull/48592) |

## Detection of weak generic types
This was reverted by [#48923](https://github.com/microsoft/TypeScript/pull/48923)

The generic weak types on constraints has caused all our packages to fail due to an error in a base package. The specific pattern that we ran into is this:

```ts
type Spec = {
    props?: string[]
}
type Class<T extends Spec> = { s: Spec }
function cls<T>(s: T):Class<T> { return null! }
```
[Playground Link 4.6](https://www.typescriptlang.org/play?ts=4.6.2&ssl=5&ssc=48&pln=1&pc=1#code/C4TwDgpgBAypDGUC8UDeAoKWpgE4HswBnAfgC4ojhcBLAOwHMBtAXXQF91RIoBhAGwCGRIgB4AKlAgAPYBDoATIrAQA+ZGkoU4ERJwBmAVzrxgNfHSjx+Y8aoAURCuICUZAcNvrUUXBGCGuJZ0hvz8AIRQ7EA) / [Playground Link 4.7](https://www.typescriptlang.org/play?ts=4.7.0-dev.20220505#code/C4TwDgpgBAypDGUC8UDeAoKWpgE4HswBnAfgC4ojhcBLAOwHMBtAXXQF91RIoBhAGwCGRIgB4AKlAgAPYBDoATIrAQA+ZGkoU4ERJwBmAVzrxgNfHSjx+Y8aoAURCuICUZAcNvrUUXBGCGuJZ0hvz8AIRQ7EA)

This was already reverted but we are implementing fixes based on the error to ensure if this is turned on in the future we will be better prepared.

## Symbol is now included in `keyof T`

This was the biggest source of breaking changes (post reverting weak generic type detection) breaking on the order of 2% of packages. We ran into this when concatenating or interpolating `keyof T` into a string:

```ts
function add<T>(p: keyof T) {
    return "prefix." + p;
}
function message<T>(p: keyof T) {
    return `prefix.${p}`;
}
```

[Playground Link 4.6](https://www.typescriptlang.org/play?ts=4.6.2#code/GYVwdgxgLglg9mABAQwCaoDwBUB8AKABwC5EBrAUwE85hEsBKRAbwChF3EAncqETpAEQFuwGAA8AdAMQBqRAQDcLAL4sWoSLASIAtuQDO+5AHNy2fMTJUadRqw5cefJAANh5UZIAkTAspdKqkA) / [Playground Link 4.7](https://www.typescriptlang.org/play?ts=4.7.0-dev.20220506#code/GYVwdgxgLglg9mABAQwCaoDwBUB8AKABwC5EBrAUwE85hEsBKRAbwChF3EAncqETpAEQFuwGAA8AdAMQBqRAQDcLAL4sWoSLASIAtuQDO+5AHNy2fMTJUadRqw5cefJAANh5UZIAkTAspdKqkA)

This new inclusion of symbol appears to be too eager in certain scenarios:

```ts
export type Context = {
    V1: { "a": string },
    V2: { "b": string },
}
function path<V extends keyof Context, K extends keyof Context[V]>(v: V, k: K)  {
    // v is treated as string
    // k is not treated as a string, even though it can only ever be "a" | "b"
    return `${v}.${k}`
}
```
[Playground Link 4.6](https://www.typescriptlang.org/play?ts=4.6.2#code/KYDwDg9gTgLgBDAnmYcDCEB2NTwLxwDeAUHGXAGoCMAXEXAEQCGDdAzjFAJaYDmcAXwA0pchQBMdQowBGrOB259BIgcWIAzAK6YAxjC5Y4YJjAAWAHgpxcwTABM2cANbBEEDeiw4QMIXABpG187Rxc3Dy9sXABtCgBdAD4ACgA3Ogp-ZzoAgEoyEnI4AHpiuFS4LidFU2B7OCZqzh5eUTJSl0qnTAh4TmBa+saGhWa+f2BUuwQzCC1eM0r4XSZMOCwAG0QbKag4GVRmBjgAH1kGNrgoYBgtKDWAAwASQlSBADoX5wEH4gEgA) / [Playground Link 4.7](https://www.typescriptlang.org/play?ts=4.7.0-dev.20220506#code/KYDwDg9gTgLgBDAnmYcDCEB2NTwLxwDeAUHGXAGoCMAXEXAEQCGDdAzjFAJaYDmcAXwA0pchQBMdQowBGrOB259BIgcWIAzAK6YAxjC5Y4YJjAAWAHgpxcwTABM2cANbBEEDeiw4QMIXABpG187Rxc3Dy9sXABtCgBdAD4ACgA3Ogp-ZzoAgEoyEnI4AHpiuFS4LidOYFNgezgmJ0UeXlEyUpdKp0wIeBq6hqbGhU5W-2BUuwQzCC1eM0r4XSZMOCwAG0QbKag4GVRmBjgAH1kGdrgoYBgtKDWAAwASQlSBADoX5wEH4gEgA)
 
## Better detection of always true conditions

This caused 5 errors in 0.5% of packages, 1 of which is definitely an error (an un-awaited promise). The others are method definitions that are checked to exist, mostly in tests so they are probably intentional.

## Changes in declaration emit 

This version changed declaration emit in several ways that we found to be benign:
1. Comments on default exports are now moved on the symbol that is exported not the default export statement 
2. Optional parameters no longer emit `| undefined` on the parameter type if it was not explicitly added (this seems like a improvement)
3. One declaration map changed for unknown reasons
4. Changes to how conditional types are inlined. This is due to [this PR](https://github.com/microsoft/TypeScript/pull/48592). This seems like an improvement but we are still investigating the impact on downstream consumers of these packages.

**Comment by DanielRosenwasser** [comment]
Thank you for reporting with the results @dragomirtitian! A few questions/comments:

> The others are method definitions that are checked to exist, mostly in tests so they are probably intentional

Have you found the experience of squelching these issues to be reasonable?

> Comments on default exports are now moved on the symbol that is exported not the default export statement

I would likely consider this a bug if you're willing to file it.

> Symbol is now included in `keyof T`

This is my least favorite unexpected breaking change because anecdotally, nobody cares about symbols. The usual mitigation is to write `string & keyof T`. How bad was this experience?

**Comment by dragomirtitian** [comment]
> > The others are method definitions that are checked to exist, mostly in tests so they are probably intentional
> 
> Have you found the experience of squelching these issues to be reasonable?

It was actually pretty straightforward. In most cases this was a redundant check (for example an assertion function called right before). 

> 
> > Comments on default exports are now moved on the symbol that is exported not the default export statement
> 
> I would likely consider this a bug if you're willing to file it.

This is a simple repro of it. 

```ts
/**
 * Comment
 */
export default {
   test: 0
}
```
[Playground Link 4.6](https://www.typescriptlang.org/play?ts=4.6.2#code/PQKhCgAIUhhB7AtogpgOwC5RMcKAeADvAE4aQAmKAZgIYCuANuQN5SSQYoDOGAXJAAM4AL5A) / [Playground Link 4.7](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220512&ssl=6&ssc=2&pln=1&pc=1#code/PQKhCgAIUhhB7AtogpgOwC5RMcKAeADvAE4aQAmKAZgIYCuANuQN5SSQYoDOGAXJAAM4AL5A)


I would hesitate to make this a bug. It's just a change, not sure if it was intentional or not. It does not affect us one way or the other, we just saw it as a difference (we generally check if any build output changes between TS versions and investigate why it did). 

We initially feared typedoc might be impacted buy this, but it turns out it wasn't including the comment on the default export in the docs anyway, so I don't see a change there either. 

The only noticeable change is that tooltips in VS Code now pick up the comment and surface it, which is an improvement.

> 
> > Symbol is now included in `keyof T`
> 
> This is my least favorite unexpected breaking change because anecdotally, nobody cares about symbols. The usual mitigation is to write `string & keyof T`. How bad was this experience?

We mostly encountered this when composing log messages, so we made a runtime change to explicitly convert using `String(key)`. We did this to avoid breaking changes in dependencies, changing .
